qa/tasks/cephfs/test_volumes.py

```diff
@@ -1650,6 +1907,234 @@ def test_subvolume_rm_with_snapshots(self):
+
+        # recreate retained subvolume using its own snapshot to clone
+        self._fs_cmd("subvolume", "snapshot", "clone", self.volname, subvolume, snapshot1, subvolume)
+
+        # check clone status
+        self._wait_for_clone_to_complete(subvolume)
+
+        # verify clone
+        self._verify_clone(subvolume, clone1)
+
+        # retain a clone
+        self._fs_cmd("subvolume", "rm", self.volname, clone1, "--retain-snapshots")
+
+        # do some IO
+        self._do_subvolume_io(subvolume, number_of_files=64)
+
+        # create a newer snapshot of subvolume
+        self._fs_cmd("subvolume", "snapshot", "create", self.volname, subvolume, snapshot2)
+
+        # remove subvolume with snapshot retention
+        self._fs_cmd("subvolume", "rm", self.volname, subvolume, "--retain-snapshots")
+
+        # clone from newer snapshot
+        self._fs_cmd("subvolume", "snapshot", "clone", self.volname, subvolume, snapshot2, clone2)
+
+        # check clone status
+        self._wait_for_clone_to_complete(clone2)
+
+        # create a snapshot on the clone
+        self._fs_cmd("subvolume", "snapshot", "create", self.volname, clone2, snapshot3)
+
+        # recreate a retained clone from another clone
+        self._fs_cmd("subvolume", "snapshot", "clone", self.volname, clone2, snapshot3, clone1)
+
+        # check clone status
+        self._wait_for_clone_to_complete(clone1)
+
+        # verify clone
+        self._verify_clone(clone2, clone1)
+
+        # remove clone with snapshot retention (no snapshots present)
+        self._fs_cmd("subvolume", "rm", self.volname, clone1, "--retain-snapshots")
+
+        # remove clone with snapshot retention (snapshots present)
+        self._fs_cmd("subvolume", "rm", self.volname, clone2, "--retain-snapshots")
+
+        ## check disallowed operations on retained clone
+        # clone-status
+        try:
+            self._fs_cmd("clone", "status", self.volname, clone2)
+        except CommandFailedError as ce:
+            self.assertEqual(ce.exitstatus, errno.EAGAIN, "invalid error code on clone status of clone with retained snapshots")
+        else:
+            self.fail("expected clone status of clone with retained snapshots to fail")
+
+        # clone-cancel
+        try:
+            self._fs_cmd("clone", "cancel", self.volname, clone2)
+        except CommandFailedError as ce:
+            self.assertEqual(ce.exitstatus, errno.EAGAIN, "invalid error code on clone cancel of clone with retained snapshots")
+        else:
+            self.fail("expected clone cancel of clone with retained snapshots to fail")
+
+        # remove snapshots (removes volumes as all are in retained state)
+        self._fs_cmd("subvolume", "snapshot", "rm", self.volname, subvolume, snapshot1)
+        self._fs_cmd("subvolume", "snapshot", "rm", self.volname, subvolume, snapshot2)
+        self._fs_cmd("subvolume", "snapshot", "rm", self.volname, clone2, snapshot3)
+
```

Comment:
You can maybe `subvolume ls` and check whether they are removed?

Reply:
Done, nice idea.